chain/index/events.go

```diff
@@ -0,0 +1,548 @@
+			return xerrors.Errorf("failed to get message id for message cid and tipset key cid: %w", err)
+		}
+		if messageID == 0 {
+			return xerrors.Errorf("message id not found for message cid %s and tipset key cid %s", em.msg.Cid(), msgTs.Key())
+		}
+
+		// Insert events for this message
+		for _, event := range em.evs {
+			addr, found := addressLookups[event.Emitter]
+			if !found {
+				var ok bool
+				addr, ok = si.idToRobustAddrFunc(ctx, event.Emitter, executionTs)
+				if !ok {
+					// not an address we will be able to match against
+					continue
+				}
+				addressLookups[event.Emitter] = addr
+			}
+
+			robustAddrbytes := addr.Bytes()
+			if addr.Protocol() != address.Delegated {
+				robustAddrbytes = nil
+			}
+
+			// Insert event into events table
+			eventResult, err := tx.Stmt(si.stmts.insertEventStmt).ExecContext(ctx, messageID, eventCount, uint64(event.Emitter), robustAddrbytes, 0)
+			if err != nil {
+				return xerrors.Errorf("failed to insert event: %w", err)
+			}
+
+			// Get the event_id of the inserted event
+			eventID, err := eventResult.LastInsertId()
+			if err != nil {
+				return xerrors.Errorf("failed to get last insert id for event: %w", err)
+			}
+
+			// Insert event entries
+			for _, entry := range event.Entries {
+				_, err := tx.Stmt(si.stmts.insertEventEntryStmt).ExecContext(ctx,
+					eventID,
+					isIndexedValue(entry.Flags),
+					[]byte{entry.Flags},
+					entry.Key,
+					entry.Codec,
+					entry.Value,
+				)
+				if err != nil {
+					return xerrors.Errorf("failed to insert event entry: %w", err)
+				}
+			}
+			eventCount++
+		}
+	}
+
+	return nil
+}
+
+func (si *SqliteIndexer) loadExecutedMessages(ctx context.Context, msgTs, rctTs *types.TipSet) ([]executedMessage, error) {
+	msgs, err := si.cs.MessagesForTipset(ctx, msgTs)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get messages for tipset: %w", err)
+	}
+
+	st := si.cs.ActorStore(ctx)
+
+	receiptsArr, err := blockadt.AsArray(st, rctTs.Blocks()[0].ParentMessageReceipts)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to load message receipts array: %w", err)
+	}
+
+	if uint64(len(msgs)) != receiptsArr.Length() {
+		return nil, xerrors.Errorf("mismatching message and receipt counts (%d msgs, %d rcts)", len(msgs), receiptsArr.Length())
+	}
+
+	ems := make([]executedMessage, len(msgs))
+
+	for i := 0; i < len(msgs); i++ {
+		ems[i].msg = msgs[i]
+
+		var rct types.MessageReceipt
+		found, err := receiptsArr.Get(uint64(i), &rct)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to load receipt %d: %w", i, err)
+		}
+		if !found {
+			return nil, xerrors.Errorf("receipt %d not found", i)
+		}
+		ems[i].rct = rct
+
+		// no events in the receipt
+		if rct.EventsRoot == nil {
+			continue
+		}
+
+		eventsArr, err := amt4.LoadAMT(ctx, st, *rct.EventsRoot, amt4.UseTreeBitWidth(types.EventAMTBitwidth))
+		if err != nil {
```

Comment:
this branch should be guarded by `ipld.IsNotFound(err)` - if it's simply a block-not-found then you can do all this recompute logic, otherwise you should just bail straight away

Reply:
@rvagg Even then -> retrying once dosen't hurt no ? Getting an error other than `ipld.IsNotFound` really should be rare -> we are logging the error and so we're not losing that information -> we're retrying once to see if things work out.

Given that this branch of getting an error that is not `ipld.IsNotFound` is rare -> the optimisation isn't worth it IMO.

Reply:
It's not really an optimisation, it could lead to confusion - say the error is a disk error with badger, you're going to get a warning and then one of the following errors, the one you get could be deeply nested from within `recomputeTipSetStateFunc()` so we'll get a big long chained error out of that, or perhaps it only comes when we `LoadAMT`, or maybe we get a read error first and then a write error next.

Not a big deal, I'd just prefer to firewall all this logic around the specific error case we care about and properly fail on anything else and not get tangled in it weird state situations.

Reply:
This is done.